uart_mcux_lpuart: Enable Asynchronous UART API.

NickolasLapp opened the pull request:
This PR enables the Ansynchronous UART API for using the MCUX drivers.
It is tested on the RT1062EVKB. In order to make this functionality
possible, the MCUX EDMA driver is also exteneded to enable the
dma_reload functionality.

Signed-off-by: Nickolas Lapp <nickolaslapp@gmail.com>

teburd commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -127,15 +163,20 @@ static void channel_irq(edma_handle_t *handle)
+		/* Clear the done flag before the callback so a user can optionally
+		 * install (reload) the next handle during the callback if desired
+		 */
+		if (transfer_done) {
```
I'm not an edma expert, but I'm wondering if there's more transfer descriptors to do in the tcd, will marking the current transfer as done like this cause the hardware to move to the next one before the callback is done? That seems to go against the prior intent here and potentially problematic.

DerekSnell replied:
This change does break my code, and DONE bit should not be cleared before the callback.  I am updating an I2S driver to use scatter/gather mode of eDMA, and also added dma_reload() to the eDMA driver (I am working now to merge in your updates before submitting my PR).  channel_irq() in this file is based on EDMA_HandleIRQ() from SDK eDMA driver, and we should be careful about diverging from the SDK driver.  The SDK driver clears the DONE bit after the callback for a specific reason.

When using scatter/gather mode, EDMA_SubmitTransfer() has some logic for the coherency check when TCD registers are updated.  That logic depends on the DONE bit set in some cases.  The I2S test case I am working on is one of these cases, and the SubmitTransfer() does not work properly with this update where the DONE bit is cleared before the callback.

DerekSnell replied:
This change does break my code, and DONE bit should not be cleared before the callback. I am updating an I2S driver to use scatter/gather mode of eDMA, and also added dma_reload() to the eDMA driver (I am working now to merge in your updates before submitting my PR). channel_irq() in this file is based on EDMA_HandleIRQ() from SDK eDMA driver, and we should be careful about diverging from the SDK driver. The SDK driver clears the DONE bit after the callback for a specific reason.

When using scatter/gather mode, EDMA_SubmitTransfer() has some logic for the coherency check when TCD registers are updated. That logic depends on the DONE bit set in some cases. The I2S test case I am working on is one of these cases, and the SubmitTransfer() does not work properly with this update where the DONE bit is cleared before the callback.

NickolasLapp replied:
In the case of the UART ASYNC driver, the callback will trigger a call to DMA reload when the currently used buffer is full, and the callback requests from the user, via: [UART_RX_BUF_REQUEST](https://docs.zephyrproject.org/latest/reference/peripherals/uart.html#c.uart_event_type.UART_RX_BUF_REQUEST)

If dma_reload is called when the done bit is set (as a result of `uart_rx_buf_rsp`) then the DMA transfer is not restarted and the ping-pong transfer stops. This problem was mitigated by clearing the DONE bit before calling into the user callback.

I'll make the change back now to clear the done bit where as done in the SDK driver, but will have to look at how to correctly modify the UART driver so that transfers continue in this case.

teburd commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -205,24 +247,56 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
+
 	switch (config->channel_direction) {
-	case MEMORY_TO_MEMORY:
+	case MEMORY_TO_MEMORY: {
```
These brackets in this switch case aren't necessary

NickolasLapp replied:
I thought uncrustify added these for me, but I will remove for now.

teburd replied:
We are moving away from uncrustify, I know the docs still mention them, but please do try the newer clang format setup if you are interested in doing auto formatting like me.  See #41307

teburd commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -205,24 +247,56 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
 	}
+	}
+
+	if (!data_size_valid(config->source_data_size)) {
```
this is a nice cleanup of this check

teburd reviewed (commented):
Overall looks ok to me, there's some stuff I'd like to see comments/changes on before my +1. The serial driver changes are large so I'm leaving it up to others to review those as I'm not as familiar with it.

mmahadevan108 commented:
@NickolasLapp Thank you for the contribution. Can you please share how this feature was tested.

DerekSnell commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -33,16 +33,42 @@ struct dma_mcux_edma_config {
+static __aligned(32) __nocache edma_tcd_t
+tcdpool[DT_INST_PROP(0, dma_channels)][CONFIG_DMA_TCD_QUEUE_SIZE];
+#else
+#warning tcdpool may be located in cacheable memory which can cause EDMA issues
```
Agreed.  using cacheable memory should not be allowed if the driver doesn't maintain cache coherency.  I think it is best to force the tcdpool in non-cacheable memory, and this should be a #error

NickolasLapp replied:
Agreed--I was concerned about breaking the ability to compile projects which had used this in the past but did not have proper cache setup. I agree an error is more appropriate here so I will make this change.

DerekSnell reviewed (changes requested):
Thank you for implementing all these features.  Clearing the eDMA DONE bit in the ISR must be done after the callback.  Let me know if I can help with your code to work with that change.  Other than that change, the rest of my feedback are comments.

dleach02 commented:
@NickolasLapp thank you for your contribution. @DerekSnell and I have been discussing the cache coherency changes you put into the UART driver when interfacing with the EDMA. I'm not sure that the proper place for those changes are in the UART driver as it seems like they should be put down into the EDMA driver. My thinking is that multiple stacks use the EDMA driver and would potentially have the same coherency issues and if we centralized the code in the edma driver all  of the other drivers would get the same benefit. Otherwise, the code you added to the UART  driver would need to be duplicated to the other drivers.

Thoughts?

NickolasLapp commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -33,16 +33,42 @@ struct dma_mcux_edma_config {
 	void (*irq_config_func)(const struct device *dev);
 };
 
+#ifdef CONFIG_DMA_MCUX_USE_DTCM_FOR_DMA_BUFFER
```
I agree this is better! Will make the change.

NickolasLapp commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -33,16 +33,42 @@ struct dma_mcux_edma_config {
 	void (*irq_config_func)(const struct device *dev);
 };
 
+#ifdef CONFIG_DMA_MCUX_USE_DTCM_FOR_DMA_BUFFER
```
Agreed, will make the change!

NickolasLapp commented:
Sorry all about the delay in getting back to your feedback. I got caught up in some other work things and just now got back to this. I appreciate the thorough review and will make the changes as requested


> @NickolasLapp thank you for your contribution.
> 
> @DerekSnell and I have been discussing the cache coherency changes you put into the UART driver when interfacing with the EDMA. I'm not sure that the proper place for those changes are in the UART driver as it seems like they should be put down into the EDMA driver. My thinking is that multiple stacks use the EDMA driver and would potentially have the same coherency issues and if we centralized the code in the edma driver all of the other drivers would get the same benefit. Otherwise, the code you added to the UART driver would need to be duplicated to the other drivers.
> 
> Thoughts?

I agree that it would be preferable to include the cache clean / cache invalidate operations in the `dma_mcux_edma.c` instead of in the uart driver. I will put up a commit to move that logic from the UART driver into the EDMA driver.

NickolasLapp commented:
> @NickolasLapp Thank you for the contribution. Can you please share how this feature was tested.

@mmahadevan108 I tested this on an RT1062EVKB. We have a zephyr project which is based closely on the RT1062EVK project in zephyr. I have 2 UART interfaces connected (lpuart0 and lpuart3) and I passed data with checksum back and forth between the two UARTs, confirming the correct data was received.

NickolasLapp commented:
@dleach02 and @DerekSnell , I was thinking a bit about this comment: 
> I'm not sure that the proper place for those changes are in the UART driver as it seems like they should be put down into the EDMA driver. My thinking is that multiple stacks use the EDMA driver and would potentially have the same coherency issues and if we centralized the code in the edma driver all of the other drivers would get the same benefit. Otherwise, the code you added to the UART driver would need to be duplicated to the other drivers.

and remembered why I had not done this in the first place. My concern is that the dcache operations have to operate on a full cache line (32 bytes for the processor I'm working on), and so we round up the length for invalidate / clean operations as appropriate to ensure we always pass a multiple of the DCACHE line length to the `DCACHE_xxx` operations.

This could be an issue, however, if the buffers passed to the EDMA module are not padded / aligned to the appropriate value. For example, if an end user did something like:

```
struct my_data {
   uint8_t data[16];
   double some_important_double;
}
```
and passed `my_data.data` as the rx buffer to the edma engine, `some_important_double` could get improperly invalidated (assuming it's only been written to cache, and the value hasn't been written back to main memory) in the call to `DCACHE_Invalidate` because we invalidate the full cache line which could include that extra value.

I think this is probably a fine tradeoff given that the alternative is the data that's received is possibly not correct because the cache operations haven't been performed after the DMA rx, but this is definitely non-obvious behavior and not documented in the zephyr DMA page (which makes no mention of aligning the DMA buffers).

I still think this functionality does belong inside the EDMA driver itself, but I'm not sure in general what the right behavior here is.

NickolasLapp commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -164,19 +166,15 @@ static void channel_irq(edma_handle_t *handle)
 		}
 
+		if (transfer_done) {
+			handle->base->CDNE = handle->channel;
```
I went back through and figured out why it was an issue to call `EDMA_SubmitTransfer()` from within the callback ISR context.

If you look at `fsl_edma.c`, considering the case of a ping-pong buffer, we end up in this block:
https://github.com/NXPmicro/mcux-sdk/blob/79cdc5799191a7411479a2617e28a191f4383276/drivers/edma/fsl_edma.c#L1176-L1207

In this block, we've correctly loaded the next DMA descriptor to point to, and now we are trying to set the ESG bit for the active DMA descriptor to enable it to continue on into the reloaded buffer. However, because we haven't cleared the DONE bit yet, writes to ESG are disabled and so it looks to the driver like the previous transfer finished (see the comment [here](https://github.com/NXPmicro/mcux-sdk/blob/main/drivers/edma/fsl_edma.c#L1209)).

I get around this in the PR that I pushed where I just delay reloading the DMA reload until we're out of ISR context in the case that `dma_reload` is called from w/in ISR, but this feels a bit ugly to me. Happy to take suggestions if you have thoughts about a better way to handle this.

NickolasLapp replied:
Sounds good! I think I've got this correctly updated to use the changes from the hal_nxp PR ([#144](https://github.com/zephyrproject-rtos/hal_nxp/pull/144/)), and to use the `HandleIRQ` function from within the driver itself instead of the custom `channel_irq`. I tested with the driver fixes on my board and all seems to work as intended.

DerekSnell commented:
> @dleach02 and @DerekSnell , I was thinking a bit about this comment:
> 
> > I'm not sure that the proper place for those changes are in the UART driver as it seems like they should be put down into the EDMA driver. My thinking is that multiple stacks use the EDMA driver and would potentially have the same coherency issues and if we centralized the code in the edma driver all of the other drivers would get the same benefit. Otherwise, the code you added to the UART driver would need to be duplicated to the other drivers.
> 
> and remembered why I had not done this in the first place. My concern is that the dcache operations have to operate on a full cache line (32 bytes for the processor I'm working on), and so we round up the length for invalidate / clean operations as appropriate to ensure we always pass a multiple of the DCACHE line length to the `DCACHE_xxx` operations.
> 
> This could be an issue, however, if the buffers passed to the EDMA module are not padded / aligned to the appropriate value. For example, if an end user did something like:
> 
> ```
> struct my_data {
>    uint8_t data[16];
>    double some_important_double;
> }
> ```
> 
> and passed `my_data.data` as the rx buffer to the edma engine, `some_important_double` could get improperly invalidated (assuming it's only been written to cache, and the value hasn't been written back to main memory) in the call to `DCACHE_Invalidate` because we invalidate the full cache line which could include that extra value.
> 
> I think this is probably a fine tradeoff given that the alternative is the data that's received is possibly not correct because the cache operations haven't been performed after the DMA rx, but this is definitely non-obvious behavior and not documented in the zephyr DMA page (which makes no mention of aligning the DMA buffers).
> 
> I still think this functionality does belong inside the EDMA driver itself, but I'm not sure in general what the right behavior here is.

Hi @NickolasLapp ,
Once again, thank you for thinking through this and raising this concern.  It is a good point, and we have been investigating how best to handle these potential use-case issues.  But we think it is best not to burden your PR with this.  We propose you move the cache handling into the eDMA driver, and simply invalidate the RX buffers and clean the TX buffers.  And then we can create a new GitHub issue to improve handling the cache in this driver.  You are welcome to create the issue, otherwise I can.  

Thank you

NickolasLapp commented:
> > @dleach02 and @DerekSnell , I was thinking a bit about this comment:
> > > I'm not sure that the proper place for those changes are in the UART driver as it seems like they should be put down into the EDMA driver. My thinking is that multiple stacks use the EDMA driver and would potentially have the same coherency issues and if we centralized the code in the edma driver all of the other drivers would get the same benefit. Otherwise, the code you added to the UART driver would need to be duplicated to the other drivers.
> > 
> > 
> > and remembered why I had not done this in the first place. My concern is that the dcache operations have to operate on a full cache line (32 bytes for the processor I'm working on), and so we round up the length for invalidate / clean operations as appropriate to ensure we always pass a multiple of the DCACHE line length to the `DCACHE_xxx` operations.
> > This could be an issue, however, if the buffers passed to the EDMA module are not padded / aligned to the appropriate value. For example, if an end user did something like:
> > ```
> > struct my_data {
> >    uint8_t data[16];
> >    double some_important_double;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and passed `my_data.data` as the rx buffer to the edma engine, `some_important_double` could get improperly invalidated (assuming it's only been written to cache, and the value hasn't been written back to main memory) in the call to `DCACHE_Invalidate` because we invalidate the full cache line which could include that extra value.
> > I think this is probably a fine tradeoff given that the alternative is the data that's received is possibly not correct because the cache operations haven't been performed after the DMA rx, but this is definitely non-obvious behavior and not documented in the zephyr DMA page (which makes no mention of aligning the DMA buffers).
> > I still think this functionality does belong inside the EDMA driver itself, but I'm not sure in general what the right behavior here is.
> 
> Hi @NickolasLapp , Once again, thank you for thinking through this and raising this concern. It is a good point, and we have been investigating how best to handle these potential use-case issues. But we think it is best not to burden your PR with this. We propose you move the cache handling into the eDMA driver, and simply invalidate the RX buffers and clean the TX buffers. And then we can create a new GitHub issue to improve handling the cache in this driver. You are welcome to create the issue, otherwise I can.
> 
> Thank you


I spent some time this morning digging into this, and the more I think about it, the more I think it's not feasible to do the DCACHE invalidate/clean operations from within the EDMA driver itself. The parameters to the callback `nxp_edma_callback` don't have enough information to know exactly when the buffers should be cleaned / invalidated, just based on the nxp driver design.

Instead, what I think makes sense (and I added a commit which implements this behavior) is that we just disallow DMA transfers to memory locations located in cacheable memory. I think this is probably the use case for 90% of users anyway, and it seems to be what's recommended in the MCUX example projects. 

Let me know what you think about this commit or if there is anything else I can do to work towards getting this PR merged.

DerekSnell commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -18,6 +18,12 @@
+#ifdef CONFIG_HAS_MCUX_CACHE
+#include <cache.h>
+#include <fsl_cache.h>
+#include <linker/linker-defs.h>
```
These two includes are no longer needed with your changes, since you removed the cache APIs from this driver.
#include <cache.h>
#include <fsl_cache.h>

Below header will not be needed if removing the check of buffers being in cacheable memory
#include <linker/linker-defs.h>

DerekSnell commented:
> I spent some time this morning digging into this, and the more I think about it, the more I think it's not feasible to do the DCACHE invalidate/clean operations from within the EDMA driver itself. The parameters to the callback `nxp_edma_callback` don't have enough information to know exactly when the buffers should be cleaned / invalidated, just based on the nxp driver design.
> 
> Instead, what I think makes sense (and I added a commit which implements this behavior) is that we just disallow DMA transfers to memory locations located in cacheable memory. I think this is probably the use case for 90% of users anyway, and it seems to be what's recommended in the MCUX example projects.
> 
> Let me know what you think about this commit or if there is anything else I can do to work towards getting this PR merged.

Hi @NickolasLapp ,
I understand the challenges with putting cache coherency checks in the eDMA driver.  But forcing DMA buffers to be in non-cacheable regions also has issues.  Zephyr's DMA APIs do not have this requirement, and forcing this in eDMA driver would change the API.  Applications can use buffers in cached memory today with these drivers if the app manages the cache coherency.  So forcing this requirement in the driver could break existing code.  And since the DMA driver is generic API that applies across all architectures, then if this requirement is added, it should get acceptance from the Zephyr community to make such a change. 

We do need to handle the cache coherency better.  But as mentioned before, we don't want to burden your PR with this topic.  You have contributed some helpful updates to the drivers we want to merge.  And I have another I2S PR waiting on yours, that also resolves some issues in a different driver.  I propose you remove the non-cache requirement in the eDMA driver to move this PR forward.  And then we can create an Enhancement Issue to track the desire to improve the cache handling in the eDMA driver.

Thanks again for all your contribution

DerekSnell commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -224,6 +229,57 @@ static int dma_mcux_edma_configure(const struct device *dev, uint32_t channel,
+			}
+		}
+	}
+#endif // CONFIG_HAS_MCUX_CACHE
```
Per [our discussion](https://github.com/zephyrproject-rtos/zephyr/pull/42750#issuecomment-1061102220), let's remove the requirement that buffers are in non-cacheable memory.

DerekSnell reviewed (changes requested):
Per [our discussion](https://github.com/zephyrproject-rtos/zephyr/pull/42750#issuecomment-1061102220), I think it is best to remove the requirement for buffers in non-cacheable memory.  Otherwise, this review has some minor tweaks.

Thanks again, @NickolasLapp !

dleach02 commented on tests/crypto/mbedtls/prj.conf:
```diff
@@ -6,3 +6,4 @@ CONFIG_MBEDTLS_TEST=y
 CONFIG_ZTEST=y
 CONFIG_TEST_USERSPACE=y
 CONFIG_MINIMAL_LIBC_RAND=y
+CONFIG_NOCACHE_MEMORY=y
```
this change doesn't seem relevant to your PR

NickolasLapp replied:
I was trying to fix the CI/CD issues (The spi loopback CI unit test was failing because there was no non-cacheable memory in which to locate the TCDPool). I've pushed a new change that I _think_ should fix this without these kconfig options though.

dleach02 commented on tests/drivers/spi/spi_loopback/prj.conf:
```diff
@@ -4,3 +4,4 @@ CONFIG_SPI=y
 CONFIG_SPI_ASYNC=y
 CONFIG_SPI_LOG_LEVEL_INF=y
 CONFIG_ZTEST=y
+CONFIG_NOCACHE_MEMORY=y
```
This doesn't seem relevant to your PR

dleach02 commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -83,8 +138,8 @@ static void mcux_lpuart_poll_out(const struct device *dev, unsigned char c)
-	while (!(LPUART_GetStatusFlags(config->base)
-		& kLPUART_TxDataRegEmptyFlag)) {
+	while (!(LPUART_GetStatusFlags(config->base) &
+		 kLPUART_TxDataRegEmptyFlag)) {
```
In the future, you should resist the desire to do whitespace changes in the file you are adding changes to. It creates noise to the reviewers.

NickolasLapp replied:
This was a remnant from the uncrustify auto-formatting (I realize this has been replaced by clang-tidy, but the documentation I read when starting to work on this indicated to format all changes with the uncrustify config that was checked into the repo).
Will keep this in mind for the future, however!

dleach02 commented:
@NickolasLapp can you address the CI issues so we can do some testing on this PR?

NickolasLapp commented:
> @NickolasLapp can you address the CI issues so we can do some testing on this PR?
> 
> Also, what testing have you done to verify the changes?

@dleach02 Working on addressing the CI issues! Just pushed a change that will hopefully fix the remaining problems.

For testing, I've validated the UART driver with and without the ASYNC API on an RT1060 EVK-B Eval Board. I have a test build which passes data back and forth from a host to the CPU with random data & checksum and validates the roundtrip data (this using the async API).
Additionally, I tested the UART console/logging shell and confirmed these worked with the async api disabled.

NickolasLapp commented on tests/drivers/spi/spi_loopback/overlay-mcux-dspi-dma.conf:
```diff
@@ -3,4 +3,4 @@ CONFIG_DMA=y
 CONFIG_DSPI_MCUX_EDMA=y
 CONFIG_MCUX_DSPI_BUFFER_SIZE=144
 CONFIG_SPI_LOOPBACK_FAST_FREQ=1000000
-CONFIG_DMA_MCUX_USE_DTCM_FOR_DMA_DESCRIPTORS=n
```
Was planning on squashing all the fixup commits into a single commit before merging to avoid the commit-guidelines rules, but leaving them as is for reviewers for now.

Have squashed and also addressed I think all of the other compliance issues if you don't mind kicking off the compliance check again @danieldegrasse . Thanks for noticing!

danieldegrasse commented:
@NickolasLapp How are you testing this implementation? I'm seeing errors with the `dma_request_channel` calls, likely because the EDMA driver initializes in the post kernel stage, and the LPUART driver initializes in the pre kernel stage, so the EDMA driver would not be ready to provide a channel when the LPUART driver requested one.

NickolasLapp commented:
> @NickolasLapp How are you testing this implementation? I'm seeing errors with the `dma_request_channel` calls, likely because the EDMA driver initializes in the post kernel stage, and the LPUART driver initializes in the pre kernel stage, so the EDMA driver would not be ready to provide a channel when the LPUART driver requested one.


@danieldegrasse 
I've got a custom build based off the RT1060 EVK board [project](https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/arm/mimxrt1060_evk). On top of that I've built a UART application that passes data back and forth to the serial port on my host machine.
There is a call to `uart_configure()` where my thread initializes, however, which could be masking this error on my end. Let me see if I can remove the uart configure call and reproduce the problem on my end.

If it is an issue of configuring at the wrong point, what is your recommendation for init priorities in the case where one driver depends on another? I think ideally serial continues to be `PRE_KERNEL` because it should be one of the first things initialized, but I'm not sure we should pull the EDMA driver into `PRE_KERNEL` as well...
Let me check what other similar projects have done.

NickolasLapp commented:
I can confirm I have the same errors if I don't call `uart_configure` within my application level code. 
I am able to resolve the error by changing the priority of the EDMA driver initialization to `PRE_KERNEL_1` (from POST_KERNEL) and changing the priority of the LPUART initialization to `PRE_KERNEL_2`. Is that okay @danieldegrasse ?

danieldegrasse commented:
@NickolasLapp that does fix the failure to get a channel, but the UART test I'm using still does not pass. I think the test that this PR should pass is `tests/drivers/uart/uart_async_api`. When I use your fix, that test outputs data, but it seems to be out of order (see the log below):
```


/%ssage 1Message 2  /* NOTE: Message 2 is a value that this test case should print, so the testcase is running */



O
.

?
```

The fix I would suggest over switching initialization order is instead statically allocating the DMA channels in the DTS, and removing calls to `dma_request_channel` The LPSPI driver does this, see [mimxrt1060_evk.dts](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts#L219) and [spi_mcux_lpspi.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/spi/spi_mcux_lpspi.c#L564). My concern with `dma_request_channel` is that it does not seem to have a way to track which channels are in use, which means you could use the same DMA channel for multiple devices.

Using static DMA channel allocation, I get the following output from the test (which seems to be progress in the right direction):

```
*** Booting Zephyr OS build zephyr-v3.0.0-946-gc7c708075c6d  ***
Running test suite uart_async_test
===================================================================
START - test_single_read_setup
 PASS - test_single_read_setup in 0.1 seconds
===================================================================
START - test_single_read
test
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:88: test_single_read: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_single_read in 0.218 seconds
===================================================================
START - test_multiple_rx_enable_setup
 PASS - test_multiple_rx_enable_setup in 0.1 seconds
===================================================================
START - test_multiple_rx_enable
E: Unable to start receive on UART.

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:136: test_multiple_rx_enable: (ret not equal to 0)
uart_rx_enable failed
 FAIL - test_multiple_rx_enable in 0.19 seconds
===================================================================
START - test_chained_read_setup
 PASS - test_chained_read_setup in 0.1 seconds
===================================================================
START - test_chained_read
E: Unable to start receive on UART.



M
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:268: test_chained_read: (k_sem_take(&rx_rdy, K_MSEC(1000)) not equal to 0)
RX_RDY timeout
 FAIL - test_chained_read in 1.31 seconds
===================================================================
START - test_double_buffer_setup
 PASS - test_double_buffer_setup in 0.1 seconds
===================================================================
START - test_double_buffer
E: Unable to start receive on UART.

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:321: test_double_buffer: (uart_rx_enable(uart_dev, double_buffer[0], sizeof(double_buffer[0]), 50 * USEC_PER_MSEC) not equal to 0)
Failed to enable receiving
 FAIL - test_double_buffer in 0.26 seconds
===================================================================
START - test_read_abort_setup
 PASS - test_read_abort_setup in 0.1 seconds
===================================================================
START - test_read_abort
E: Unable to start receive on UART.

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:404: test_read_abort: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_read_abort in 0.121 seconds
===================================================================
START - test_chained_write_setup
 PASS - test_chained_write_setup in 0.1 seconds
===================================================================
START - test_chained_write
E: Unable to start receive on UART.
Message 1Message 2
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:618: test_chained_write: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_chained_write in 0.122 seconds
===================================================================
START - test_long_buffers_setup
 PASS - test_long_buffers_setup in 0.1 seconds
===================================================================
START - test_long_buffers
E: Unable to start receive on UART.

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:690: test_long_buffers: (k_sem_take(&rx_rdy, K_MSEC(200)) not equal to 0)
RX_RDY timeout
 FAIL - test_long_buffers in 0.264 seconds
===================================================================
START - test_write_abort_setup
 PASS - test_write_abort_setup in 0.1 seconds
===================================================================
START - test_write_abort
E: Unable to start receive on UART.

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:476: test_write_abort: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_write_abort in 0.121 seconds
===================================================================
START - test_forever_timeout_setup
 PASS - test_forever_timeout_setup in 0.1 seconds
===================================================================
START - test_forever_timeout
E: Unable to start receive on UART.
/%



O
.

?
/
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:552: test_forever_timeout: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_forever_timeout in 3.121 seconds
===================================================================
Test suite uart_async_test failed.
===================================================================
PROJECT EXECUTION FAILED
```

I'm interested to see if you can reproduce these failures on your end. Also, when you update this PR, could you add the appropriate `dmas` and `dma-names` property to the lpuart peripherals? It's a simple thing to fix in testing, but should be done before this PR goes upstream.

NickolasLapp commented:
@danieldegrasse I pushed a change with those fixes. It looks like `tests/drivers/uart/uart_async_api` are passing for me (at least locally on ubuntu specifying the RT1060EVK as the test platform. Let me know if that all looks good to you!

danieldegrasse commented:
@NickolasLapp I'm still seeing some errors- see my log below:
```
*** Booting Zephyr OS build zephyr-v3.0.0-947-g1214245f617e  ***
Running test suite uart_async_test
===================================================================
START - test_single_read_setup
 PASS - test_single_read_setup in 0.1 seconds
===================================================================
START - test_single_read
test
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:88: test_single_read: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_single_read in 0.218 seconds
===================================================================
START - test_multiple_rx_enable_setup
 PASS - test_multiple_rx_enable_setup in 0.1 seconds
===================================================================
START - test_multiple_rx_enable
E: Unable to start receive on UART.

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:136: test_multiple_rx_enable: (ret not equal to 0)
uart_rx_enable failed
 FAIL - test_multiple_rx_enable in 0.19 seconds
===================================================================
START - test_chained_read_setup
 PASS - test_chained_read_setup in 0.1 seconds
===================================================================
START - test_chained_read
E: Unable to start receive on UART.

        ]
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:268: test_chained_read: (k_sem_take(&rx_rdy, K_MSEC(1000)) not equal to 0)
RX_RDY timeout
 FAIL - test_chained_read in 1.31 seconds
===================================================================
START - test_double_buffer_setup
 PASS - test_double_buffer_setup in 0.1 seconds
===================================================================
START - test_double_buffer
E: Unable to start receive on UART.

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:321: test_double_buffer: (uart_rx_enable(uart_dev, double_buffer[0], sizeof(double_buffer[0]), 50 * USEC_PER_MSEC) not equal to 0)
Failed to enable receiving
 FAIL - test_double_buffer in 0.26 seconds
===================================================================
START - test_read_abort_setup
 PASS - test_read_abort_setup in 0.1 seconds
===================================================================
START - test_read_abort
E: Unable to start receive on UART.
`
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:404: test_read_abort: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_read_abort in 0.121 seconds
===================================================================
START - test_chained_write_setup
 PASS - test_chained_write_setup in 0.1 seconds
===================================================================
START - test_chained_write
E: Unable to start receive on UART.
Message 1Message 2
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:618: test_chained_write: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_chained_write in 0.122 seconds
===================================================================
START - test_long_buffers_setup
 PASS - test_long_buffers_setup in 0.1 seconds
===================================================================
START - test_long_buffers
E: Unable to start receive on UART.

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:690: test_long_buffers: (k_sem_take(&rx_rdy, K_MSEC(200)) not equal to 0)
RX_RDY timeout
 FAIL - test_long_buffers in 0.264 seconds
===================================================================
START - test_write_abort_setup
 PASS - test_write_abort_setup in 0.1 seconds
===================================================================
START - test_write_abort
E: Unable to start receive on UART.
!X
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:476: test_write_abort: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_write_abort in 0.121 seconds
===================================================================
START - test_forever_timeout_setup
 PASS - test_forever_timeout_setup in 0.1 seconds
===================================================================
START - test_forever_timeout
E: Unable to start receive on UART.
II``V`!Xß&`ï:`¤:`X±D`%<`%<`
                           M

.
        ?
st
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/uart/uart_async_api/src/test_uart_async.c:552: test_forever_timeout: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_forever_timeout in 3.122 seconds
===================================================================
Test suite uart_async_test failed.
===================================================================
PROJECT EXECUTION FAILED
```

This test seems to need the UART RX and TX shorted, based on the checks it makes. Are you shorting the RX and TX lines for a console?

The other option would be to use another UART device for the test (see `test_uart.h`) in the test. If we used LPUART3 as the test device and added a board overlay in the `boards` directory of this test to enable it, the test should pass with pins 1 and 2 of J22 shorted.

NickolasLapp commented:
@danieldegrasse  could you please send me the exact command you're using to test so I can see if I can reproduce on my end? I was running: `./scripts/twister -j 12  -p mimxrt1060_evk -T tests/drivers/uart/uart_async_api`

danieldegrasse commented:
@NickolasLapp twister will only build the test, unless you explicitly tell it to run it on a device (and provide a serial port to use). Personally, I typically build and flash the test directly- the `PROJECT EXECUTION SUCCESSFUL` string is usually what twister is looking for in the console output. This is the set of commands i used:

```
west build -p always -b mimxrt1060_evk tests/drivers/uart/uart_async_api
west flash --reset-after-load
```

NickolasLapp commented:
Thanks for the tip @danieldegrasse , I've got this building and flashing now for the RT1060 EVK-B after a bit of modification for the different board (I'm a bit new to this workflow since we primarily develop on mac-os, so getting familiar with the development environment from Ubuntu).
My next question is how did you get the above output after flashing the async uart test? Is this available over the RTT logger, or somewhere else?

danieldegrasse commented:
@NickolasLapp If you modify `test_uart.h` to use LPUART3 for the test device, and add a board overlay enabling LPUART3 the test should be able to pass. I think you'll need to loopback LPUART3 by shorting pins 1 and 2 of J22 on the EVK. The other potential issue is that `dma_request_channel` may return the same channel for both LPUART devices

NickolasLapp commented:
@danieldegrasse That works great for me. I imagine it should be relatively quick to merge through the edma channel / pinmux settings, and then I can rebase as soon as that lands. Please just post here and let me know when the PR is available with these changes and I can rebase as appropriate.

danieldegrasse commented:
@NickolasLapp I've figured out a solution for the RT11xx series boards- moving buffers to noncacheable OCRAM results in the test passing. I've opened #43885 to enable that functionality, and provide appropriate pinmuxes for all RT series boards that should be tested with the uart async api. I'll update you once that PR merges, and we can proceed from there.

danieldegrasse commented:
@NickolasLapp FYI- #43885 is merged, so this PR should be ready for testing once it is rebased against the latest main.

NickolasLapp commented:
Thanks for the update @danieldegrasse .
I rebased and fixed the merge conflicts, then re-ran the test suite as well as my local board tests and confirmed everything passed:
```*** Booting Zephyr OS build zephyr-v3.0.0-2356-g565a646f28f0  ***
Running test suite uart_async_test
===================================================================
START - test_single_read_setup
 PASS - test_single_read_setup in 0.1 seconds
===================================================================
START - test_single_read
E: No buffers to release from RX DMA!
 PASS - test_single_read in 0.356 seconds
===================================================================
START - test_multiple_rx_enable_setup
 PASS - test_multiple_rx_enable_setup in 0.1 seconds
===================================================================
START - test_multiple_rx_enable
E: No buffers to release from RX DMA!
E: No buffers to release from RX DMA!
 PASS - test_multiple_rx_enable in 0.710 seconds
===================================================================
START - test_chained_read_setup
 PASS - test_chained_read_setup in 0.1 seconds
===================================================================
START - test_chained_read
E: No buffers to release from RX DMA!
 PASS - test_chained_read in 0.221 seconds
===================================================================
START - test_double_buffer_setup
 PASS - test_double_buffer_setup in 0.1 seconds
===================================================================
START - test_double_buffer
 PASS - test_double_buffer in 3.403 seconds
===================================================================
START - test_read_abort_setup
 PASS - test_read_abort_setup in 0.1 seconds
===================================================================
START - test_read_abort
 PASS - test_read_abort in 1.110 seconds
===================================================================
START - test_chained_write_setup
 PASS - test_chained_write_setup in 0.1 seconds
===================================================================
START - test_chained_write
E: No buffers to release from RX DMA!
E: No buffers to release from RX DMA!
 PASS - test_chained_write in 0.9 seconds
===================================================================
START - test_long_buffers_setup
 PASS - test_long_buffers_setup in 0.1 seconds
===================================================================
START - test_long_buffers
 PASS - test_long_buffers in 0.152 seconds
===================================================================
START - test_write_abort_setup
 PASS - test_write_abort_setup in 0.1 seconds
===================================================================
START - test_write_abort
 PASS - test_write_abort in 0.102 seconds
===================================================================
START - test_forever_timeout_setup
 PASS - test_forever_timeout_setup in 0.1 seconds
===================================================================
START - test_forever_timeout
E: No buffers to release from RX DMA!
E: No buffers to release from RX DMA!
 PASS - test_forever_timeout in 3.5 seconds
===================================================================
Test suite uart_async_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```
Looking forward to getting this merged finally! :)

danieldegrasse commented:
@NickolasLapp I've tested this PR on the RT1170, RT1064, RT1050, and RT1024 EVKs, and it passes on all of them. Can you squash your commits appropriately so that they will pass CI?

NickolasLapp commented:
Squashed and pushed @danieldegrasse ! Thanks for testing

danieldegrasse commented:
@NickolasLapp just a heads up: you can check for issues like trailing whitespace using the `checkpatch` script, see [here](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style). Just don't want you to have to wait on the compliance check report from CI every time.

Also, can you split the changes to the EDMA driver into a separate commit? This will make traceability easier, since the changes within the EDMA driver are intended to properly support scatter/gather mode, which is required for this PR but a separate enhancement from enabling the UART async API itself.

NickolasLapp committed:
dma_mcux_edma: Fixup Scatter-Gather EDMA Mode

This PR fixes up the Scatter-Gather EDMA mode for the MCUX EDMA Driver,
as well as enabling the dma reload feature for the same EDMA Driver.

Signed-off-by: Nickolas Lapp <nickolaslapp@gmail.com>

NickolasLapp committed:
uart_mcux_lpuart: Enable Asynchronous UART API.

This PR enables the Ansynchronous UART API for using the MCUX drivers.
It is tested on the RT1062EVKB.

Signed-off-by: Nickolas Lapp <nickolaslapp@gmail.com>

NickolasLapp commented:
Got it. 🤞 that it runs now and all tests pass. Checkpatch passes locally:
```
- total: 0 errors, 0 warnings, 1235 lines checked
Your patch has no obvious style problems and is ready for submission.
```

DerekSnell reviewed (approved):
Looks great to me.  Thanks for all your hard work, @NickolasLapp !
